src/main/java/hudson/remoting/Launcher.java

```diff
@@ -244,6 +268,90 @@ public void run() throws Exception {
+            } catch (CertificateException e) {
+                throw new IllegalStateException("Java platform specification mandates support for X.509", e);
+            }
+            KeyStore keyStore = Engine.getCacertsKeyStore();
+            // load the keystore
+            keyStore.load(null, null);
+            int i = 0;
+            for (String certOrAtFilename : candidateCertificates) {
+                certOrAtFilename = certOrAtFilename.trim();
+                byte[] cert;
+                if (certOrAtFilename.startsWith("@")) {
+                    File file = new File(certOrAtFilename.substring(1));
+                    long length;
+                    if (file.isFile()
+                            && (length = file.length()) < 65536
+                            && length > "-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----".length()) {
+                        FileInputStream fis = null;
+                        try {
+                            // we do basic size validation, if there are x509 certificates that have a PEM encoding
+                            // larger
+                            // than 64kb we can revisit the upper bound.
+                            cert = new byte[(int) length];
+                            fis = new FileInputStream(file);
+                            int read = fis.read(cert);
+                            if (cert.length != read) {
+                                LOGGER.log(Level.WARNING, "Only read {0} bytes from {1}, expected to read {2}",
+                                        new Object[]{read, file, cert.length});
+                                // skip it
+                                continue;
+                            }
+                        } catch (IOException e) {
+                            LOGGER.log(Level.WARNING, "Could not read certificate from " + file, e);
+                            continue;
+                        }
+                    } else {
+                        if (file.isFile()) {
+                            LOGGER.log(Level.WARNING,
+                                    "Could not read certificate from {0}. File size is not within " +
+                                            "the expected range for a PEM encoded X.509 certificate",
+                                    file.getAbsolutePath());
+                        } else {
+                            LOGGER.log(Level.WARNING, "Could not read certificate from {0}. File not found",
+                                    file.getAbsolutePath());
+                        }
+                        continue;
+                    }
+                } else {
+                    try {
+                        cert = certOrAtFilename.getBytes("US-ASCII");
+                    } catch (UnsupportedEncodingException e) {
+                        LOGGER.log(Level.WARNING, "Could not parse certificate " + certOrAtFilename, e);
```

Comment:
US-ASCII is a standard encoding in Java IIRC. Maybe it makes sense to throw IllegalStateException like you do for X.509